jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ContractionHierarchyBidirectionalDijkstra.java

```diff
@@ -0,0 +1,377 @@
+/*
+ * (C) Copyright 2019-2019, by Semen Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.shortestpath;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.graph.EdgeReversedGraph;
+import org.jgrapht.graph.GraphWalk;
+import org.jgrapht.graph.MaskSubgraph;
+import org.jheaps.AddressableHeap;
+import org.jheaps.tree.PairingHeap;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static org.jgrapht.alg.shortestpath.BidirectionalDijkstraShortestPath.DijkstraSearchFrontier;
+import static org.jgrapht.alg.shortestpath.ContractionHierarchy.ContractionEdge;
+import static org.jgrapht.alg.shortestpath.ContractionHierarchy.ContractionVertex;
+
+/**
+ * Implementation of the hierarchical query algorithm based on the bidirectional Dijkstra search.
+ * It is designed to operate on contracted graphs which makes it fast for road networks (sparse graphs
+ * with low average outdegree of vertices).
+ *
+ * <p>
+ * The query algorithm is originally described the article: Robert Geisberger, Peter Sanders, Dominik Schultes,
+ * and Daniel Delling. 2008. Contraction hierarchies: faster and simpler hierarchical routing in road networks.
+ * In Proceedings of the 7th international conference on Experimental algorithms (WEA'08), Catherine C. McGeoch (Ed.).
+ * Springer-Verlag, Berlin, Heidelberg, 319-333.
+ *
+ * <p>
+ * During contraction graph is divided into 2 parts which are called upward and downward graphs. Both parts have all
+ * vertices of the original graph. The upward graph ($G_{&#92;uparrow}$) contains only those edges which source
+ * has lower level than the target and vice versa for the downward graph ($G_{\downarrow}$).
+ *
+ * <p>
+ * For the shortest path query from $s$ to $t$, a modified bidirectional Dijkstra shortest path search is
+ * performed. The forward search from $s$ operates on $G_{&#92;uparrow}$ and the backward search from $t$ -
+ * on the $G_{\downarrow}$. In each direction only the edges of the corresponding part of the graph are
+ * considered. Both searches eventually meet at the vertex $v$, which has the highest order in the shortest
+ * path from $s$ to $t$. Whenever a node is settled in one direction that is already settled in the other
```

Comment:
Please, explain what does it mean to "settle" a node in this case. It's a bit unclear from the context.